File: 2019/src/main/java/aoc/Day01.java
package aoc;

import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;

public class Day01 {
    private long[] parse(String input) {
        String[] lines = input.replaceAll("\r\n", "\n").split("\n");
        long[] result = new long[lines.length];

        for (int i = 0; i < lines.length; i++) {
            result[i] = Long.parseLong(lines[i]);
        }

        return result;
    }

    private long calculateFuel(long mass) {
        return mass / 3 - 2;
    }

    public long part1(String input) {
        long total = 0;

        for (long mass : parse(input)) {
            total += calculateFuel(mass);
        }

        return total;
    }

    public long part2(String input) {
        long total = 0;

        for (long mass : parse(input)) {
            long fuel = calculateFuel(mass);
            while (fuel > 0) {
                total += fuel;
                fuel = calculateFuel(fuel);
            }
        }

        return total;
    }

    public static void main(String[] args) throws Exception {
        Day01 day = new Day01();
        String filepath = "src/main/resources/day01.txt";
        String input = Files.readString(Paths.get(filepath), StandardCharsets.UTF_8);

        long startPart1 = System.currentTimeMillis();
        long resultPart1 = day.part1(input);
        long endPart1 = System.currentTimeMillis();
        System.out.println("Part 1: " + resultPart1 + " (" + (endPart1 - startPart1) + " ms)");

        long startPart2 = System.currentTimeMillis();
        long resultPart2 = day.part2(input);
        long endPart2 = System.currentTimeMillis();
        System.out.println("Part 2: " + resultPart2 + " (" + (endPart2 - startPart2) + " ms)");
    }
}

File: 2020/aoc2020/src/main/kotlin/Day01.kt
class Day01 {

    private val target = 2020

    fun part1(input: String): Int = input
        .lines()
        .map { it.toInt() }
        .let { numbers ->
            numbers
                .flatMap { a -> numbers.map { b -> Pair(a, b) } }
                .find { (a, b) -> a + b == target }
                ?.let { (a, b) -> a * b }
        } ?: 0

    fun part2(input: String): Int {
        val numbers = input.lines().map { it.toInt() }
        val sortedNumbers = numbers.sorted()

        for (i in 0..<sortedNumbers.size - 2) {
            var left = i + 1
            var right = sortedNumbers.size - 1

            while (left < right) {
                val sum = sortedNumbers[i] + sortedNumbers[left] + sortedNumbers[right]

                when {
                    sum < target -> left++
                    sum > target -> right--
                    else -> return sortedNumbers[i] * sortedNumbers[left] * sortedNumbers[right]
                }
            }
        }

        return 0
    }
}

fun main() {
    val day = Day01()
    val input = readText("day01.txt")

    val startPart1 = System.currentTimeMillis()
    val answerPart1 = day.part1(input)
    val endPart1 = System.currentTimeMillis()
    println("Part 1: $answerPart1 (${endPart1 - startPart1} ms)")

    val startPart2 = System.currentTimeMillis()
    val answerPart2 = day.part2(input)
    val endPart2 = System.currentTimeMillis()
    println("Part 2: $answerPart2 (${endPart2 - startPart2} ms)")
}
